### PR TITLE
startree hll config and aggregation function in query processing

### PR DESCRIPTION
**Pull request by victor2100**
1. Add hll config in startree metadata.
2. Implement fasthll aggregation function under old and new aggregation function schemes.
3. Allow fasthll in user query to utilize the startree index.
4. Pass SegmentMetadata into AggregationFunctionContext to give hll function enough info.
5. Implement fetch string value inside SingleValueBlockCache.

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+  @Override
+  public List<HyperLogLog> combine(List<HyperLogLog> aggregationResultList, CombineLevel combineLevel) {
+    if ((aggregationResultList == null) || aggregationResultList.isEmpty()) {
+      return null;
+    }
+    HyperLogLog hllResult = aggregationResultList.get(0);
+    for (int i = 1; i < aggregationResultList.size(); ++i) {
+      try {
+        hllResult.addAll(aggregationResultList.get(i));
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java]: can aggregationResultList have null values ?

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    return aggregationResult0;
+  }
+
+  @Override
+  public Long reduce(List<HyperLogLog> combinedResultList) {
+    if ((combinedResultList == null) || combinedResultList.isEmpty()) {
+      return 0L;
+    }
+    HyperLogLog reducedResult = combinedResultList.get(0);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java]: This part can be moved to common method and shared with combine() method

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java [hunk 3]
```diff
@@ -88,6 +94,29 @@ public static ColumnMetadata fromPropertiesConfiguration(String column, Properti
     }
     builder.setPaddingCharacter(paddingCharacter);
 
+    // derived metric
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: comment will be more useful if you indicate what metadata property it's reading to check for derived type

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java [hunk 4]
```diff
@@ -328,17 +397,16 @@ public char getPaddingCharacter() {
     return paddingCharacter;
   }
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: add a method for isDerivedMetricType()

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java (hunk 1 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java]: This is the old version aggregation function, it just mimic the structure of combine of old distinctcounthll and other functions to have a pre-check of the null value. I'm not sure it's absolutely necessary, but it's always safe to do that.

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java (hunk 2 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java]: I'll fix that.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java (hunk 4 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: We have getDerivedMetricType(), from which outside methods can get the type info. And if in the future, there exists other derived types, outside methods can add a new "case" clause in switch for handling that type, which seems clear.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java (hunk 3 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: fixed

File: pinot-common/src/main/java/com/linkedin/pinot/common/response/broker/GroupByResult.java [hunk 5]
```diff
@@ -50,4 +50,9 @@ public void setValue(Serializable value) {
   public void setGroup(List<String> group) {
     _group = group;
   }
+
+  @Override
+  public String toString() {
+    return _group + ": " + _value.toString();
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/response/broker/GroupByResult.java]: The value can be very large. We shouldn't output full value in toString()

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/DefaultAggregationExecutor.java [hunk 6]
```diff
@@ -129,12 +133,18 @@ private void aggregateColumn(AggregationFunctionContext aggrFuncContext, Aggrega
       case AggregationFunctionFactory.DISTINCTCOUNT_AGGREGATION_FUNCTION:
       case AggregationFunctionFactory.DISTINCTCOUNTHLL_AGGREGATION_FUNCTION:
         aggregationFunction.aggregate(length, resultHolder,
-            _singleValueBlockCache.getHashCodeArrayForColumn(aggrColumn));
+            (Object) _singleValueBlockCache.getHashCodeArrayForColumn(aggrColumn));
+        break;
+
+      case AggregationFunctionFactory.FASTHLL_AGGREGATION_FUNCTION:
+        String derivedColumn = _segmentMetadata.getStarTreeMetadata().getDerivedHllColumnFromOrigin(aggrColumn);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/DefaultAggregationExecutor.java]: One of the review comments from Friday was to do this without startTree Metadata if the hLL columns are provided to us. You can change that as another commit but evaluate if fastHLL without starTree provides any benefit over distinctCountHLL

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunctionFactory.java [hunk 7]
```diff
@@ -69,6 +73,9 @@ public static AggregationFunction getAggregationFunction(String functionName) {
       case DISTINCTCOUNTHLL_AGGREGATION_FUNCTION:
         return new DistinctCountHLLAggregationFunction();
 
+      case FASTHLL_AGGREGATION_FUNCTION:
+        return new FastHllAggregationFunction(segmentMetadata.getStarTreeMetadata().getHllLog2m());
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunctionFactory.java]: Can getStarTreeMetadata() throw NPE ? If yes, protect from that

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @param combinedResult
+   * @return
+   */
+  @Override
+  public Double reduce(List<Object> combinedResult) {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java]: Is this todo ? 

How do you combine results from multiple segments ?

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/DistinctCountHLLAggregationFunction.java [hunk 9]
```diff
@@ -180,15 +165,7 @@ public Long reduce(List<HyperLogLog> combinedResultList) {
...
-    HyperLogLog reducedResult = combinedResultList.get(0);
-    for (int i = 1; i < combinedResultList.size(); ++i) {
-      try {
-        reducedResult.addAll(combinedResultList.get(i));
-      } catch (CardinalityMergeException e) {
-        LOGGER.error("Caught exception while merging Cardinality using HyperLogLog", e);
-        Utils.rethrowException(e);
-      }
-    }
+    HyperLogLog reducedResult = HllUtil.mergeHLLResultsToFirstInList(combinedResultList);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/DistinctCountHLLAggregationFunction.java]: Is this shared with fastHLL ? Add a comment to FastHLL class reduce method that reduce is implemented here

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java [hunk 10]
```diff
@@ -88,6 +94,30 @@ public static ColumnMetadata fromPropertiesConfiguration(String column, Properti
...
+    if (derivedMetricType != null) {
+      switch (derivedMetricType) {
+        case HLL:
+          try {
+            final int hllLog2m = config.getInt(V1Constants.MetadataKeys.StarTree.STAR_TREE_HLL_LOG2M);
+            builder.setFieldSize(HllUtil.getHllFieldSizeFromLog2m(hllLog2m));
+            final String originColumnName = config.getString(getKeyFor(column, ORIGIN_COLUMN));
+            builder.setOriginColumnName(originColumnName);
+          } catch (RuntimeException e) {
+            LOGGER.error("HLL derived column without fieldSize or originColumnName.");
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: log column name

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java [hunk 11]
```diff
@@ -88,6 +94,30 @@ public static ColumnMetadata fromPropertiesConfiguration(String column, Properti
...
+    // DERIVED_METRIC_TYPE property is used to check whether this field is derived or not
+    // ORIGIN_COLUMN property is used to indicate the origin field of this derived metric
+    String typeStr = config.getString(getKeyFor(column, DERIVED_METRIC_TYPE), null);
+    DerivedMetricType derivedMetricType = (typeStr == null)? null : DerivedMetricType.valueOf(typeStr.toUpperCase());
+
+    if (derivedMetricType != null) {
+      switch (derivedMetricType) {
+        case HLL:
+          try {
+            final int hllLog2m = config.getInt(V1Constants.MetadataKeys.StarTree.STAR_TREE_HLL_LOG2M);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: handle missing value for key V1Constants.MetadataKeys.StarTree.STAR_TREE_HLL_LOG2M

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java [hunk 12]
```diff
@@ -88,6 +94,30 @@ public static ColumnMetadata fromPropertiesConfiguration(String column, Properti
...
+            builder.setFieldSize(HllUtil.getHllFieldSizeFromLog2m(hllLog2m));
+            final String originColumnName = config.getString(getKeyFor(column, ORIGIN_COLUMN));
+            builder.setOriginColumnName(originColumnName);
+          } catch (RuntimeException e) {
+            LOGGER.error("HLL derived column without fieldSize or originColumnName.");
+            throw e;
+          }
+          break;
+        default:
+          throw new IllegalArgumentException(derivedMetricType + " type is not supported in building column metadata.");
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: add column name and segment/table information if available

File: pinot-core/src/main/java/com/linkedin/pinot/core/startree/OffHeapStarTreeBuilder.java [hunk 13]
```diff
@@ -139,7 +139,7 @@ public void init(StarTreeBuilderConfig builderConfig) throws Exception {
     if (outDir == null) {
       outDir = new File(System.getProperty("java.io.tmpdir"), V1Constants.STAR_TREE_INDEX_DIR + "_" + DateTime.now());
     }
-    LOG.debug("Index output directory:{}", outDir);
+    LOG.info("Index output directory:{}", outDir);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/startree/OffHeapStarTreeBuilder.java]: Any reason to change log level here? If this is for local testing, please revert it back

File: pinot-core/src/test/java/com/linkedin/pinot/core/startree/hll/SegmentWithHllIndexCreateHelper.java [hunk 14]
```diff
@@ -15,144 +15,152 @@
...
+  /**
+   * must call this to clean up
+   */
+  public void cleanTempDir() {
+    if (INDEX_DIR != null) {
+      FileUtils.deleteQuietly(INDEX_DIR);
+    }
+  }
+
+  private static void printSchema(Schema schema) {
```
Comment (antumbde) [inline pinot-core/src/test/java/com/linkedin/pinot/core/startree/hll/SegmentWithHllIndexCreateHelper.java]: logger.debug

File: pinot-core/src/test/java/com/linkedin/pinot/core/startree/hll/TestHllIndexCreation.java [hunk 15]
```diff
@@ -119,6 +120,21 @@ public void testColumnStatsWithStarTree() {
       for (String name : helper.getSchema().getColumnNames()) {
         LOGGER.debug("* " + name + ": " + driver.getColumnStatisticsCollector(name).getCardinality());
       }
+      LOGGER.info("Loading ...");
```
Comment (antumbde) [inline pinot-core/src/test/java/com/linkedin/pinot/core/startree/hll/TestHllIndexCreation.java]: debug log all

File: pinot-core/src/test/java/com/linkedin/pinot/queries/HllIndexSentinelTest.java [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+  @BeforeClass
+  public void setup() throws Exception {
+    setupTableManager();
+
+    // Setup Segment
+    helper = new SegmentWithHllIndexCreateHelper(tableName, AVRO_DATA, timeColumnName, timeUnit);
+    SegmentIndexCreationDriver driver = helper.build(true, hllConfig);
+    File segmentFile = helper.getSegmentDirectory();
+    segmentName = helper.getSegmentName();
+    LOGGER.info("************************** Segment Directory: " + segmentFile.getAbsolutePath());
```
Comment (antumbde) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/HllIndexSentinelTest.java]: debug log all

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+   * Performs 'fasthll' aggregation on the input array.
+   *
+   * {@inheritDoc}
+   *
+   * @param length
+   * @param resultHolder
+   * @param valueArray
+   */
+  @Override
+  public void aggregate(int length, AggregationResultHolder resultHolder, Object... valueArray) {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java]: Add unit tests

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+/**
+ * Class to implement the old 'fasthll' aggregation function,
+ * only combine and reduce related methods are used,
+ * calling other methods will throw UnsupportedOperationException.
+ *
+ * fasthll takes advantage of pre-aggregated results for fast distinct count estimation
+ *
+ * TODO: fully deprecate and remove this class when new aggregation routine is finished.
+ */
+public class FastHllAggregationFunction implements AggregationFunction<HyperLogLog, Long> {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java]: unit tests

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java [hunk 19]
```diff
@@ -88,6 +94,30 @@ public static ColumnMetadata fromPropertiesConfiguration(String column, Properti
     }
     builder.setPaddingCharacter(paddingCharacter);
 
+    // DERIVED_METRIC_TYPE property is used to check whether this field is derived or not
+    // ORIGIN_COLUMN property is used to indicate the origin field of this derived metric
+    String typeStr = config.getString(getKeyFor(column, DERIVED_METRIC_TYPE), null);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: Unit tests

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java [hunk 20]
```diff
@@ -291,6 +297,31 @@ private void init() {
   private void initStarTreeMetadata() {
     _starTreeMetadata = new StarTreeMetadata();
 
+    // Set hll
+    _starTreeMetadata.setEnableHll(
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java]: unit tests

**Comment by antumbde** [comment]
@victor2100 This is almost 1500 lines of change. Break this PR into smaller more manageable classes. You can submit generation, metadata and aggregation related independent units separately.

Add unit tests for newer functionality. Let's discuss if something will require lot of infrastructure for unit testing.

File: pinot-common/src/main/java/com/linkedin/pinot/common/response/broker/GroupByResult.java (hunk 5 above)
Comment (victor2100) [inline pinot-common/src/main/java/com/linkedin/pinot/common/response/broker/GroupByResult.java]: The toString should not be used in any part of the real production code. It's only used for related test codes. The calling should have a sense of how much the output is and whether to use it or not. We can surely attach a warning in the comment of this function.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/DefaultAggregationExecutor.java (hunk 6 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/DefaultAggregationExecutor.java]: I will update the logic in another PR.

File: pinot-core/src/main/java/com/linkedin/pinot/core/startree/OffHeapStarTreeBuilder.java (hunk 13 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/startree/OffHeapStarTreeBuilder.java]: This is made intentionally. I believe it should be at "info" level since it does not contain any of the debug message, e.g. intermediate results. And similar messages in other classes are also at info level.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunctionFactory.java (hunk 7 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunctionFactory.java]: I will pull hll config in metadata out of StarTreeMetadata into segment metadata in next PR, so it will resolve this issue.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java (hunk 8 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java]: The reduce part of this new aggregation function is never got called. I think the transition of old aggregation API to new API is not fully completed. Currently the reduce is supported by old API and I think it should work fine for multiple segments. (That's part of the reason we need to keep the old API in system)

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/DistinctCountHLLAggregationFunction.java (hunk 9 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/DistinctCountHLLAggregationFunction.java]: No, fasthll is not shared with distinctcounthll. But old and new fasthll API is shared in combine/reduce part. I add the comments there.

File: pinot-core/src/test/java/com/linkedin/pinot/core/startree/hll/TestHllIndexCreation.java (hunk 15 above)
Comment (victor2100) [inline pinot-core/src/test/java/com/linkedin/pinot/core/startree/hll/TestHllIndexCreation.java]: fixed

File: pinot-core/src/test/java/com/linkedin/pinot/queries/HllIndexSentinelTest.java (hunk 16 above)
Comment (victor2100) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/HllIndexSentinelTest.java]: fixed

File: pinot-core/src/test/java/com/linkedin/pinot/core/startree/hll/SegmentWithHllIndexCreateHelper.java (hunk 14 above)
Comment (victor2100) [inline pinot-core/src/test/java/com/linkedin/pinot/core/startree/hll/SegmentWithHllIndexCreateHelper.java]: Still I think these are info level information. e.g. directory, column names and so on. Not the debug level information, e.g. intermediate results and so on.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java (hunk 11 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: When there is no missing value for the key, the config.getInt(key) throws an exception, which is immediately caught by my code in several lines below.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java (hunk 12 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: column name is available, added it to the exception description.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java (hunk 17 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/FastHllAggregationFunction.java]: added

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java (hunk 18 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/function/FastHllAggregationFunction.java]: added

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java (hunk 10 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: done

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java (hunk 19 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/ColumnMetadata.java]: added in ColumnMetadataTest

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java (hunk 20 above)
Comment (victor2100) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/SegmentMetadataImpl.java]: added in ColumnMetadataTest

File: pinot-core/src/test/java/com/linkedin/pinot/queries/HllIndexSentinelTest.java [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+        new MetricsRegistry()));
+  }
+
+  @AfterClass
+  public void tearDown() {
+    helper.cleanTempDir();
+  }
+
+  @Test
+  public void testFastHllNoGroupBy() throws Exception {
```
Comment (antumbde) [inline pinot-core/src/test/java/com/linkedin/pinot/queries/HllIndexSentinelTest.java]: Add a test that executes query for multiple segments. You can do that as a separate PR

**Comment by antumbde** [comment]
lgtm
will after travis gives green signal
